doc/rst/language/evolution.rst

```diff
@@ -8,6 +8,232 @@ capture significant language changes that have the possibility of breaking
+
+.. _readme-evolution.class-memory-management:
+
+class memory management
+***********************
+
+Before 1.18, if ``C`` is a class type, a variable of type ``C`` needed
+to be deleted in order to prevent a memory leak. For example:
+
+.. code-block:: chapel
+
+  class C {
+    var x: int;
+  }
+  proc main() {
+    var instance: C = new C(1);
+    delete instance;
+  }
+
+Version 1.18 introduced four memory management strategies that form part
+of a class type and are used with `new` expressions:
+
+``owned C``
+  ``owned`` classes will be deleted automatically when the ``owned``
+  variable goes out of scope, but only one ``owned`` variable can refer to
+  the instance at a time.
+  Such instances can be created with ``new owned C()``.
+
+``shared C``
+  ``shared`` classes will be deleted when all of the ``shared`` variables
+  referring to the instance go out of scope.
+  Such instances can be created with ``new shared C()``.
+
+``borrowed C``
+  refers to a class instance that has a lifetime managed by
+  another variable.
+  Values of type ``borrowed C`` can be created with ``new borrowed
+  C()``, by coercion from the other class ``C`` types, or by explicitly
+  calling the ``.borrow()`` method on one of the other class ``C``
+  types.
+  ``new borrowed C()`` creates a temporary instance that will automatically
+  be deleted at the end of the current block.
+
+``unmanaged C``
+  the instance must have `delete` called on it explicitly to
+  reclaim its memory.
+  Such instances can be created with ``new unmanaged C()``.
+
+Further note that the default is ``borrowed``, that is:
+
+``C``
+  is now the same as ``borrowed C``
+
+``new C()``
+  is now the same as ``new borrowed C()``
+
+Now, back to the example above. There are several ways to translate this
+program.
+
+First, the most semantically similar option is to replace uses of ``C``
+with ``unmanaged C``:
+
+.. code-block:: chapel
+
+  class C {
+    var x: int;
+  }
+  proc main() {
+    var instance: unmanaged C = new unmanaged C(1);
+    delete instance;
+  }
+
+Using ``unmanaged`` allows a Chapel programmer to opt in to manually
+managing the memory of the instances.
+
+A reasonable alternative would be to translate the program to use
+``owned C``:
+
+.. code-block:: chapel
+
+  class C {
+    var x: int;
+  }
+  proc main() {
+    var instance: owned C = new owned C(1);
+    // instance will now be automatically deleted at the end of this block
+  }
+
+If the program does not explicitly use ``owned C``, it can rely on
+``new C()`` being equivalent to ``new borrowed C()``:
+
+.. code-block:: chapel
+
+  class C {
+    var x: int;
+  }
+  proc main() {
+    var instance: C = new C(1);
+    // above line could be written as
+    //   var instance: borrowed C = new borrowed C(1);
+    // or as
+    //   var instance: borrowed C = (new owned C(1)).borrow();
+
+    // instance will now be automatically deleted at the end of this block
+  }
```

Comment:
I think a mention of the "Class New" section of the spec (if that's the right section) would be useful. Up to you.